File: python_src/cartridge_sdk/ui/status_bar.py
"""Top status bar widget."""

from __future__ import annotations

from typing import TYPE_CHECKING

import pygame

from cartridge_sdk.screen import _get_font
from cartridge_sdk.ui.widget import Widget
from cartridge_sdk.ui.wifi import WifiStatus, get_wifi_status

if TYPE_CHECKING:
    from cartridge_sdk.theme import Theme


class StatusBar(Widget):
    """Top bar showing app title, WiFi status, and optional status text."""

    def __init__(self, title: str = "") -> None:
        self.title = title
        self.right_text: str = ""
        self.right_color: tuple | None = None
        self._wifi_status: WifiStatus = get_wifi_status()
        self._wifi_timer: float = 0.0

    def update(self, dt: float) -> None:
        """Call each frame to refresh WiFi status periodically."""
        self._wifi_timer += dt
        if self._wifi_timer >= 10.0:
            self._wifi_timer = 0.0
            self._wifi_status = get_wifi_status()

    def draw(self, surface: pygame.Surface, rect: pygame.Rect, theme: Theme) -> None:
        # Background
        pygame.draw.rect(surface, theme.bg_header, rect)

        pad = 12
        font = _get_font("mono_bold", 16)

        # Title (left)
        title_surf = font.render(self.title, True, theme.text)
        surface.blit(title_surf, (rect.x + pad, rect.y + (rect.height - title_surf.get_height()) // 2))

        # Right text (status)
        right_edge = rect.right - pad
        if self.right_text:
            right_font = _get_font("mono", 13)
            color = self.right_color or theme.text_dim
            right_surf = right_font.render(self.right_text, True, color)
            rx = right_edge - right_surf.get_width()
            ry = rect.y + (rect.height - right_surf.get_height()) // 2
            surface.blit(right_surf, (rx, ry))
            right_edge = rx - 14

        # WiFi indicator (to the left of right_text)
        wifi = self._wifi_status
        wifi_font = _get_font("mono", 11)
        cy = rect.y + rect.height // 2

        if wifi.connected:
            if wifi.signal_strength > 50:
                dot_color = theme.positive
            elif wifi.signal_strength > 20:
                dot_color = theme.text_warning
            else:
                dot_color = theme.negative
            wifi_label = wifi_font.render("WiFi", True, theme.text_dim)
            wifi_w = 8 + 4 + wifi_label.get_width()
            wx = right_edge - wifi_w
            pygame.draw.circle(surface, dot_color, (wx + 4, cy), 3)
            surface.blit(wifi_label, (wx + 12, cy - wifi_label.get_height() // 2))
        else:
            wifi_label = wifi_font.render("No WiFi", True, theme.text_dim)
            wx = right_edge - wifi_label.get_width()
            surface.blit(wifi_label, (wx, cy - wifi_label.get_height() // 2))

        # Bottom border
        pygame.draw.line(surface, theme.border, (rect.x, rect.bottom - 1), (rect.right, rect.bottom - 1))

File: crates/cartridge-core/src/ui/status_bar.rs
//! Top status bar widget -- 36px height, gradient bg, title left, WiFi dot + clock right.

use sdl2::pixels::Color;
use sdl2::rect::Rect;

use crate::screen::Screen;

/// WiFi connection status for display in the status bar.
pub struct WifiStatus {
    pub connected: bool,
    pub signal_strength: i32, // 0-100, -1 = unknown
}

impl Default for WifiStatus {
    fn default() -> Self {
        Self {
            connected: false,
            signal_strength: -1,
        }
    }
}

/// Top bar showing app title, WiFi status dot, and clock/right text.
///
/// Height: 36px. Draws a gradient background from `header_gradient_top` to
/// `header_gradient_bottom`, a 1px accent line at the very top (y=0), the title
/// on the left, and WiFi dot + right text on the right.
pub struct StatusBar {
    pub title: String,
    pub right_text: String,
    pub right_color: Option<Color>,
    pub wifi: WifiStatus,
}

impl StatusBar {
    pub fn new(title: &str) -> Self {
        Self {
            title: title.to_string(),
            right_text: String::new(),
            right_color: None,
            wifi: WifiStatus::default(),
        }
    }

    /// Draw the status bar within `rect` (expected 36px tall).
    pub fn draw(&self, screen: &mut Screen, rect: Rect) {
        let theme = screen.theme;
        let pad = 12;

        // Gradient background
        screen.draw_gradient_rect(rect, theme.header_gradient_top, theme.header_gradient_bottom);

        // 1px accent-colored line at the very top of the bar ("brand bar")
        screen.draw_line(
            (rect.x(), rect.y()),
            (rect.x() + rect.width() as i32 - 1, rect.y()),
            Some(theme.accent),
            1,
        );

        // Title (left, bold, 16px, vertically centered)
        let title_h = screen.get_line_height(16, true);
        let title_y = rect.y() + (rect.height() as i32 - title_h as i32) / 2;
        screen.draw_text(
            &self.title,
            rect.x() + pad,
            title_y,
            Some(theme.text),
            16,
            true,
            None,
        );

        // Right edge cursor -- we build from right to left
        let mut right_edge = rect.x() + rect.width() as i32 - pad;

        // Right text (clock, status, etc.) -- 13px, dim
        if !self.right_text.is_empty() {
            let color = self.right_color.unwrap_or(theme.text_dim);
            let text_w = screen.get_text_width(&self.right_text, 13, false);
            let text_h = screen.get_line_height(13, false);
            let rx = right_edge - text_w as i32;
            let ry = rect.y() + (rect.height() as i32 - text_h as i32) / 2;
            screen.draw_text(&self.right_text, rx, ry, Some(color), 13, false, None);
            right_edge = rx - 14;
        }

        // WiFi indicator dot + label
        let cy = rect.y() + rect.height() as i32 / 2;
        if self.wifi.connected {
            let dot_color = if self.wifi.signal_strength > 50 {
                theme.positive
            } else if self.wifi.signal_strength > 20 {
                theme.text_warning
            } else {
                theme.negative
            };

            let label = "WiFi";
            let label_w = screen.get_text_width(label, 11, false);
            let label_h = screen.get_line_height(11, false);
            let wifi_w = 8 + 4 + label_w as i32; // dot_diameter + gap + label
            let wx = right_edge - wifi_w;

            // Dot (radius 3)
            screen.draw_circle(wx + 4, cy, 3, dot_color);

            // "WiFi" label
            screen.draw_text(
                label,
                wx + 12,
                cy - label_h as i32 / 2,
                Some(theme.text_dim),
                11,
                false,
                None,
            );
        } else {
            let label = "No WiFi";
            let label_w = screen.get_text_width(label, 11, false);
            let label_h = screen.get_line_height(11, false);
            let wx = right_edge - label_w as i32;
            screen.draw_text(
                label,
                wx,
                cy - label_h as i32 / 2,
                Some(theme.text_dim),
                11,
                false,
                None,
            );
        }

        // Bottom border line
        screen.draw_line(
            (rect.x(), rect.y() + rect.height() as i32 - 1),
            (
                rect.x() + rect.width() as i32 - 1,
                rect.y() + rect.height() as i32 - 1,
            ),
            Some(theme.border),
            1,
        );
    }
}
